joomla/joomla-cms
# [4.0] Post Installation: Fix Messages Notification Count

Pull Request for Issue #33311 
~And I think this also improves the PR for #29950~

Edit: The caching part implemented in this PR is not the most optimal way to do it. Thanks, @PhilETaylor for helping me understand this. So for now, I'm striking out the caching part and I'm keeping this PR open as it still fixes Issue #33311
Wherever I've mentioned Caching in this PR, it is only valid when Caching has been enabled from the global configuration settings.

### Summary of Changes

#### Before the PR
- To display the count of messages, the Post Installation Messages Module called the `getItems()` function from Post Installation Messages Component's Messages Model
- This returned an array which was contained an array of all messages **filtered by eid** (extension id, defaults to 212 for Joomla CMS) which meant that this only accounted for a portion of the total messages
- The count of this filtered array was used to display the messages count

#### After the PR
- Now, the Post Installation Messages Module calls the newly added `getItemsCount()` function that only returns the `count` (in integer) of all messages from the database **not filtered by** the default extension id (eid). Hence this count reflects all the array elements and not just the ones corresponding to Joomla CMS (for example)
- Only this count is passed to the module thus reducing the memory and performance overhead of passing array
- This count is directly used to display the message count (avoids the need of using count() function on the array)

You might observe that I've only queried 4 columns in this query. 
```
$db->quoteName('language_extension'),
$db->quoteName('language_client_id'),
$db->quoteName('condition_file'),
$db->quoteName('condition_method'),
```
This is done because the `onProcessList()` function that filters the language and extra messages uses these 4 fields. Check the link below
https://github.com/joomla/joomla-cms/blob/4d071ffe2d9c18f03b0c374795a1bd8a8ca824d6/administrator/components/com_postinstall/src/Model/MessagesModel.php#L299

![PR](https://user-images.githubusercontent.com/53610833/116115481-10353000-a6d8-11eb-9819-349b1a89880d.png)


### Testing Instructions

Firstly, make sure you have enabled Caching is enabled in Admin Panel -> Global Configuration -> System

![image](https://user-images.githubusercontent.com/53610833/116116407-bf720700-a6d8-11eb-9fe8-2184790399c8.png)


1. Run this SQL to add a fake post installation message for the Atum Template (change the db prefix!) - this replicates you installing a new extension that inserted a new post install message.
```
INSERT INTO `jos_postinstall_messages` (`postinstall_message_id`, `extension_id`, `title_key`, `description_key`, `action_key`, `language_extension`, `language_client_id`, `type`, `action_file`, `action`, `condition_file`, `condition_method`, `version_introduced`, `enabled`)
VALUES
	(99, 210, 'TEST', 'TEST', '', 'com_cpanel', 1, 'message', '', '', '', '', '4.0.0', 1);
```
Reload the admin home page.

2. Now try hiding a few messages from Joomla CMS and Atum Administrator Panel each and ensure that the count always reflects the correct value which is the sum of all messages inclusive of each extension id.

### Actual result BEFORE applying this Pull Request
![image](https://user-images.githubusercontent.com/53610833/116117471-ea108f80-a6d9-11eb-99ec-c9d004e2f0c2.png)

Note: 24 Queries after 2 refreshes (2 refreshes are done to ensure that caching is successful)


### Expected result AFTER applying this Pull Request

![image](https://user-images.githubusercontent.com/53610833/116117631-16c4a700-a6da-11eb-9206-166b6484860e.png)

Note: Still 24 queries (caching works) after 2 refreshes and the count displayed (4) is correct as it corresponds to 3 of Joomla CMS + 1 of Atum Administrator Template

You might encounter a bug while deleting and resetting messages here. Please add https://github.com/joomla/joomla-cms/pull/33338 to ensure that you can hide and reset messages easily

https://user-images.githubusercontent.com/53610833/116118563-1547ae80-a6db-11eb-8c04-56aba38f22b5.mp4

Also fixed dashboard badge: (Found in Admin Panel -> System)

![image](https://user-images.githubusercontent.com/53610833/116125234-dd446980-a6e2-11eb-835d-dcbcbf240d94.png)



### Documentation Changes Required
I am not sure but if documentation for the Post Installation messages component exists for Joomla 4 then this new function `getItemsCount()` should be mentioned under the Messages Model and the rephrased variable (integer) `$messagesCount` must be mentioned in Post Installation Modules Documentation
